librbd: API: async open and close

trociny opened the pull request:
Fixes: #14264
Signed-off-by: Mykola Golub <mgolub@mirantis.com>

trociny commented:
@dillaman Do you think it is a good approach?

Right now it leaks ImageCtx in the case open failed. I just don't know where it could be safely deleted. Doing this in C_OpenComplete::finish() does not work, because it is executed in the context of ictx->op_work_queue.

dillaman commented:
On failure, you could just have the AioCompletion delete the ImageCtx upon release since that should be coming from the caller's context.

trociny commented:
@dillaman 

> On failure, you could just have the AioCompletion delete the ImageCtx upon release since that should be coming from the caller's context.

Indeed. Thanks! Will try this.

trociny commented:
Rebased after fixing ictx leak on open failure and making close before open async too.

@dillaman it is still called rbd_aio_open, I will rename to rbd_open_async if my comment about chosing aio_open don't change your mind.

dillaman commented:
Thinking about it, it would probably be good to have an async close method as well since it can involve RADOS calls.

I would also add a new #define to the librbd.h header so apps know that async open / close are available (like discard, etc).

trociny commented:
Rebased after adding async close. Also LIBRBD_SUPPORTS_AIO_OPEN has been defined in librbd.h.

BTW, I see when a new feature is added, either a new LIBRBD_SUPPORTS_XXX define is added, or LIBRBD_VERSION is bumped. Is there some rule what is preferred?

wido commented:
Great! I'll try to give it a try. Master is however failing to build with my during a RocksDB so I'm currently stuck on that.

dillaman commented:
@trociny I think the original goal was for LIBRBD_VERSION to be bumped whenever a new API method was added -- but it looks like in practice that rarely happens and any important API additions have been flagged with their own #define

dillaman commented on src/librbd/librbd.cc:
```diff
@@ -69,6 +69,66 @@ librbd::AioCompletion* get_aio_completion(librbd::RBD::AioCompletion *comp) {
+    comp->init_time(ictx, librbd::AIO_TYPE_CLOSE);
+    comp->get();
+  }
+  virtual void finish(int r) {
```
Similar w/ aio_open, aio_close will need to delete the ImageCtx from the thread context of the caller (on success and failure)

dillaman commented on src/tracing/librbd.tp:
```diff
@@ -172,6 +172,32 @@ TRACEPOINT_EVENT(librbd, open_image_exit,
     )
 )
 
+TRACEPOINT_EVENT(librbd, aio_open_image_enter,
```
Double-check to see if rbd-replay-prep needs to be updated to handle this new event-type since it explicitly is searching for open/close events.

trociny commented:
@wido 

> Great! I'll try to give it a try. Master is however failing to build with my during a RocksDB so I'm currently stuck on that.

I am not sure it is your case but I am recalling I saw an advise to `make clean` in src/rocksdb directory.

ifed01 commented:
@wido,
Another option to override build failure that definitely works for me is to make a fresh git repo clone and execute "run-make-check.sh"

trociny commented:
Rebased after fixing  ImageCtx on close.

@dillaman I will test rbd-replay-prep (it will require some time, it is something new for me, and in the first attempt I failed to generate any librbd events :-)

dillaman commented:
@trociny You'll need to add "rbd tracing = true" to your config file to generate LTTng-UST events.

theanalyst commented:
> Great! I'll try to give it a try. Master is however failing to build with my during a RocksDB so I'm currently stuck on that.

`configure --with-librocksdb-static` may help?

trociny commented:
I have updated rbd-replay to support async open/close, it is not tested though yet, because I still have issues using rbd-replay. I added "rbd tracing = true" and was able to generate events and process them with rbd-replay-prep, but when I try to replay the generated file rbd-replay hangs in is_versioned_replay() spinning on reading the replay file. This looks like if the file is corrupted. The behavior is the same on both master and my branch. As this is my first experience with rbd-replay I might just be doing something wrong (I use this howto http://docs.ceph.com/docs/master/rbd/rbd-replay/ ). I am going to investigate this further tomorrow.

dillaman commented:
Odd -- I'll be interested to hear what the issue is with rbd-replay.  Your troubles did remind me to create a long overdue tracker ticket for creating test cases for rbd-replay.

jdurgin commented:
Does the updated rbd-replay work after your fix for EOF handling?

trociny commented:
rbd-replay basically works but still there is one important thing missed: after aio_open() the replayer should block on any further operation to this image until open is completed. I think I can fix this extending Worker::get_image() to check if open is still pending and wait for it to complete if it is. It will require some work though.

Another much simpler approach is to just treat async open/close as sync operations in the replayer (or in rbd-replay-prep). If it is ok depends how it is going to be used.

dillaman commented:
I think it's fine to treat an async open / close as a sync open / close for the purposes of rbd-replay if it's easier -- it's not like you should expect any other ops until it completes anyway.

trociny commented:
Well, if we consider the scenario the aio_open is implemented for: open a thousand of images asynchronously, then check their stats and close, it would replay much more realistically if I make it async.

Still taking I am not sure someone will use rbd-replay to test such a workload in near future, I am going to modify my patch making AioOpenImageAction::perform() and AioCloseImageAction::perform() do exactly the same what OpenImageAction::perform() and CloseImageAction::perform() do, leaving the TODO note that it might be made asynchronously.

Do you agree with this plan or may be you think there is no need in adding Aio{Open,Close}ImageAction at all in this case, and just use {Open,Close}ImageAction?

dillaman commented:
Actually, if you create new Aio{Open/Close} actions, it should result in its completion being tracked.  IO is dependent on predecessor completions -- and since aio_open re-uses AioCompletion, I think by default all IO will need to wait until the aio_open completion is processed.

trociny commented:
rbd-replay has been updated making AioOpenImageAction::perform() and AioCloseImageAction::perform() sync (doing exactly the same what OpenImageAction::perform() and CloseImageAction::perform() do). I still have plans to make it async in future (another PR).

> Actually, if you create new Aio{Open/Close} actions, it should result in its completion being tracked. IO is dependent on predecessor completions -- and since aio_open re-uses AioCompletion, I think by default all IO will need to wait until the aio_open completion is processed.

Not sure I got your point here... But the issue is not only the order of completions, the issue is that to initiate (not complete) any other operation with the image after open I need to get the image context. I have a prototype how it could be done but it is still work in progress: https://github.com/trociny/ceph/commit/0b8db4b8a9d9b16fb57d6b959a8c163216c858c0

jdurgin commented:
looks good, adding to a branch to run through the test suite
